lib/bolt/catalog.rb

```diff
@@ -65,23 +68,45 @@ def compile_catalog(request)
       target = request['target']
       pdb_client = Bolt::PuppetDB::Client.new(Bolt::PuppetDB::Config.new(request['pdb_config']))
       options = request['puppet_config'] || {}
+
       with_puppet_settings(request['hiera_config']) do
         Puppet[:rich_data] = true
         Puppet[:node_name_value] = target['name']
         Puppet::Pal.in_tmp_environment('bolt_catalog',
-                                       modulepath: request["modulepath"] || [],
-                                       facts: target["facts"] || {},
-                                       variables: target["variables"] || {}) do |pal|
+                                       modulepath: request['modulepath'] || [],
+                                       facts: target['facts'] || {},
+                                       variables: target['variables'] || {}) do |pal|
+          inv = if request['future']
+                  Bolt::ApplyInventory.new(request['config'])
+                else
+                  setup_inventory(request['inventory'])
+                end
           Puppet.override(bolt_pdb_client: pdb_client,
-                          bolt_inventory: setup_inventory(request['inventory'])) do
+                          bolt_inventory: inv) do
             Puppet.lookup(:pal_current_node).trusted_data = target['trusted']
             pal.with_catalog_compiler do |compiler|
+              if request['future']
+                # This needs to happen inside the catalog compiler so loaders are initialized for loading
+                vars = Puppet::Pops::Serialization::FromDataConverter.convert(request['plan_vars'])
+                pal.send(:add_variables, compiler.send(:topscope), vars.merge(target['vars']))
+              end
+
               # Configure language strictness in the CatalogCompiler. We want Bolt to be able
               # to compile most Puppet 4+ manifests, so we default to allowing deprecated functions.
               Puppet[:strict] = options['strict'] || :warning
               Puppet[:strict_variables] = options['strict_variables'] || false
               ast = Puppet::Pops::Serialization::FromDataConverter.convert(pal_main)
-              compiler.evaluate(ast)
+              # CODEREVIEW: Is there a better way to catch this?
+              begin
+                compiler.evaluate(ast)
+              rescue Puppet::PreformattedError => e
+                if e.message =~ /Initializer for class Bolt::ApplyTarget does not match the attributes of Target/
```

Comment:
We can make a `self.from_asserted_args` method on `ApplyTarget` that _does_ match the attributes of `Target` (so it will be called instead of `initialize`) and have it raise the error we want. We will also want `self.from_asserted_hash` to handle the hash case.

Reply:
I thought about that, but was worried that that would dredge up the conversation about why `from_asserted_*` doesn't work...I think that's a more reliable way of raising this error though.